src/test/java/LinkedAvlTreeTest.java

```diff
@@ -0,0 +1,43 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class LinkedAvlTreeTest {
+
+    private LinkedAvlTree classUnderTest;
+    private LinkedAvlTree preState;
+
+    @BeforeEach
+    void createCountryCatalogues() {
+        classUnderTest = new LinkedAvlTree();
+        preState = new LinkedAvlTree();
+    }
+
+    @Nested
+    class WhenEmpty {
+
+        @Test
+        void size_empty_returnsZero() {
+            assertThat(classUnderTest.size())
+                    .isEqualTo(0);
+        }
+
+        void isEmpty_empty_returnsTrue() {
+            assertThat(classUnderTest.isEmpty())
+                    .isTrue();
+        }
```

Comment:
FYI, these are atypical asserts when you're using assertj. AssertJ provides loads of assertions you can use on the class under test -- you should almost always prefer to use those when possible. For example here,

```java
assertThat(classUnderTest).hasSize(0);
assertThat(classUnderTest).isEmpty();
```

HOWEVER. These assertions could come from [Iterable asserts](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractIterableAssert.html#method.summary), which would use the `iterator` to do the work, which you do not have implemented. Further, those asserts would not actually test the `isEmpty` and `size` methods of the class.

TLDR: these asserts are good enough, but ensure you're looking at what assertions are available to do on the actual class under test first. Run tests locally with coverage to confirm their behaviour to yourself, and read a lot of docs. [Core asserts guide](https://assertj.github.io/doc/#assertj-core-assertions-guide)

Reply:
perogies

Reply:
> HOWEVER. These assertions could come from [Iterable asserts](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractIterableAssert.html#method.summary), which would use the iterator to do the work, which you do not have implemented. Further, those asserts would not actually test the isEmpty and size methods of the class.

They do, but nevertheless, point stands.